### PR TITLE
[BUG]Fix bugs in pruner

### PR DESCRIPTION
**Pull request by HIT-cwh**
## Motivation

Fix bugs in ``StructurePruner``.
1. Fix the bug of grouping errors caused by shared modules
2. Support trace Dilated Conv2d
3. Fix the bug that the modules whose ``requires_grad`` is False can not be traced
4. Modify the forward method in a correct way.

### CODE REVIEW

File: mmrazor/models/pruners/structure_pruning.py [hunk 1]
```diff
@@ -98,9 +111,17 @@ def prepare_from_supernet(self, supernet):
 
         # record the visited module name during trace path
         visited = dict()
+        # shared modules such as shared detection head in RetinaNet
+        self.shared_module = []
```
Comment (pppppM) [inline mmrazor/models/pruners/structure_pruning.py:115]: Add more detailed doc-strings to explain what is `shared_module` and why there need to record `shared_module`

File: mmrazor/models/pruners/structure_pruning.py [hunk 2]
```diff
@@ -656,7 +699,7 @@ def conv_backward_parser(self, grad_fn, module2name, var2module, cur_path,
         name = module2name[module]
         parent = grad_fn.next_functions[0][0]
         cur_path.append(name)
-        if visited[name]:
+        if visited[name] and name not in self.shared_module:
```
Comment (pppppM) [inline mmrazor/models/pruners/structure_pruning.py:702]: Explain why there needs `name not in self.shared_module`

File: mmrazor/models/pruners/structure_pruning.py [hunk 3]
```diff
@@ -693,7 +736,7 @@ def linear_backward_parser(self, grad_fn, module2name, var2module,
         parent = grad_fn.next_functions[1][0]
 
         cur_path.append(name)
-        if visited[name]:
+        if visited[name] and name not in self.shared_module:
```
Comment (pppppM) [inline mmrazor/models/pruners/structure_pruning.py:739]: Explain why there needs `name not in self.shared_module`

File: mmrazor/models/pruners/structure_pruning.py [hunk 4]
```diff
@@ -722,7 +765,8 @@ def concat_backward_parser(self, grad_fn, module2name, var2module,
         concat_id = '_'.join([str(id(p)) for p in parents])
         name = f'concat_{concat_id}'
         cur_path.append(name)
-        if name in visited and visited[name]:
+        if (name in visited and visited[name]
+                and name not in self.shared_module):
```
Comment (pppppM) [inline mmrazor/models/pruners/structure_pruning.py:769]: Explain why there needs `name not in self.shared_module`

File: mmrazor/models/pruners/structure_pruning.py [hunk 5]
```diff
@@ -722,7 +765,8 @@ def concat_backward_parser(self, grad_fn, module2name, var2module,
         concat_id = '_'.join([str(id(p)) for p in parents])
         name = f'concat_{concat_id}'
         cur_path.append(name)
-        if name in visited and visited[name]:
+        if (name in visited and visited[name]
```
Comment (pppppM) [inline mmrazor/models/pruners/structure_pruning.py:768]: Explain why there needs `and visited[name]`

File: mmrazor/models/pruners/structure_pruning.py [hunk 6]
```diff
@@ -89,6 +88,20 @@ def __init__(self, except_start_keys=['head.fc']):
...
+        shared modules.
+
+        Args:
+            module (:obj:`torch.nn.Module`): The module to register hook.
+            inputs (tuple): The input of the module.
+            outputs (tuple): The output of the module.
+        """
+        module.cnt += 1
+        if module.cnt == 2:
+            self.shared_module.append(self.module2name[module])
```
Comment (pppppM) [inline mmrazor/models/pruners/structure_pruning.py:103]: Add unittest to cover this line

**Review by pppppM — CHANGES_REQUESTED** [review]

File: mmrazor/models/pruners/structure_pruning.py [hunk 7]
```diff
@@ -98,9 +112,17 @@ def prepare_from_supernet(self, supernet):
...
         # record the visited module name during trace path
         visited = dict()
+        # shared modules such as shared detection head in RetinaNet
+        self.shared_module = []
 
         for name, module in supernet.model.named_modules():
             if hasattr(module, 'weight'):
+                # trace shared modules
+                module.cnt = 0
+                handle = module.register_forward_hook(
```
[the remark this answers is not in the document]
Comment (HIT-cwh) [reply mmrazor/models/pruners/structure_pruning.py:121]: Fixed, thx

File: mmrazor/models/pruners/structure_pruning.py [hunk 8]
```diff
@@ -98,9 +112,17 @@ def prepare_from_supernet(self, supernet):
...
+                    self.trace_shared_module_hook)
+                module.handle = handle
```
[the remark this answers is not in the document]
Comment (HIT-cwh) [reply mmrazor/models/pruners/structure_pruning.py:123]: Fixed, thx

File: mmrazor/models/pruners/structure_pruning.py [hunk 9]
```diff
@@ -89,6 +88,20 @@ def __init__(self, except_start_keys=['head.fc']):
...
+        """Trace shared modules. Modules such as the detection head in
+        RetinaNet which are visited more than once during :func:`forward` are
+        shared modules.
+
+        Args:
+            module (:obj:`torch.nn.Module`): The module to register hook.
+            inputs (tuple): The input of the module.
+            outputs (tuple): The output of the module.
+        """
+        module.cnt += 1
```
Comment (pppppM) [inline mmrazor/models/pruners/structure_pruning.py:101]: `module.cnt` -> `module.is_shared(bool)`?

**Review by pppppM — APPROVED** [review]

**Pull request merged**